File: test/core/models/route_error_test.dart
import 'dart:async';
import 'dart:io';

import 'package:aruku/core/models/route_error.dart';
import 'package:aruku/core/services/route_service.dart';
import 'package:aruku/l10n/app_localizations.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;

void main() {
  late AppLocalizations l10n;

  setUpAll(() async {
    l10n = await AppLocalizations.delegate.load(const Locale('ja'));
  });

  group('classifyRouteError', () {
    test('NO_ORIGIN は noLocation', () {
      expect(
        classifyRouteError(const RouteException('NO_ORIGIN')),
        RouteErrorKind.noLocation,
      );
    });

    test('NO_DESTINATION は noDestination', () {
      expect(
        classifyRouteError(const RouteException('NO_DESTINATION')),
        RouteErrorKind.noDestination,
      );
    });

    test('ZERO_RESULTS / NOT_FOUND は noResults', () {
      expect(
        classifyRouteError(const RouteException('ZERO_RESULTS')),
        RouteErrorKind.noResults,
      );
      expect(
        classifyRouteError(const RouteException('NOT_FOUND')),
        RouteErrorKind.noResults,
      );
    });

    test('HTTP エラー / 通信系例外は network', () {
      expect(
        classifyRouteError(const RouteException('HTTP 500')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(const RouteException('UNKNOWN_ERROR')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(const RouteException('UNKNOWN')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(const SocketException('failed')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(http.ClientException('reset')),
        RouteErrorKind.network,
      );
    });

    test('タイムアウトは network と別種別にする (#300)', () {
      // 上流 Transit API の遅延（正常9〜11秒・裾30秒超）は「通信断」ではない。
      // network に丸めると「通信状況を確認して再試行」と案内してしまい、電波が
      // 正常なユーザーを的外れな導線へ送る。#300 の切り分けでは画面表示から
      // App Check 拒否・レート制限・上流遅延を区別できないこと自体が障害になった。
      expect(
        classifyRouteError(const RouteException('TIMEOUT')),
        RouteErrorKind.timeout,
      );
      expect(
        classifyRouteError(TimeoutException('slow')),
        RouteErrorKind.timeout,
      );
    });

    test('タイムアウト以外の通信系は network のまま (#300)', () {
      // 分離の副作用で通信断まで timeout へ寄せていないことの反証。
      expect(
        classifyRouteError(const SocketException('failed')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(http.ClientException('reset')),
        RouteErrorKind.network,
      );
      expect(
        classifyRouteError(const RouteException('HTTP 500')),
        RouteErrorKind.network,
      );
    });

    test('NO_PROXY や未知の API status は unknown', () {
      expect(
        classifyRouteError(const RouteException('NO_PROXY')),
        RouteErrorKind.unknown,
      );
      expect(
        classifyRouteError(const RouteException('REQUEST_DENIED')),
        RouteErrorKind.unknown,
      );
      expect(classifyRouteError(Exception('boom')), RouteErrorKind.unknown);
    });
  });

  group('routeErrorView', () {
    test('各種別にタイトル・説明・主導線が定義される', () {
      for (final kind in RouteErrorKind.values) {
        final view = routeErrorView(l10n, kind);
        expect(view.kind, kind);
        expect(view.title, isNotEmpty);
        expect(view.description, isNotEmpty);
      }
    });

    test('noResults / noDestination は主導線が条件変更', () {
      expect(
        routeErrorView(l10n, RouteErrorKind.noResults).primaryRecovery,
        RouteRecovery.changeConditions,
      );
      expect(
        routeErrorView(l10n, RouteErrorKind.noDestination).primaryRecovery,
        RouteRecovery.changeConditions,
      );
    });

    test('タイムアウトは network と別の文言で、主導線は再試行 (#300)', () {
      final timeout = routeErrorView(l10n, RouteErrorKind.timeout);
      final network = routeErrorView(l10n, RouteErrorKind.network);

      expect(timeout.title, isNot(network.title));
      expect(timeout.description, isNot(network.description));
      // 上流が遅いだけで経路自体は存在し得るので、導線は条件変更でなく再試行。
      expect(timeout.primaryRecovery, RouteRecovery.retry);
    });

    test('network / noLocation / unknown は主導線が再試行', () {
      expect(
        routeErrorView(l10n, RouteErrorKind.network).primaryRecovery,
        RouteRecovery.retry,
      );
      expect(
        routeErrorView(l10n, RouteErrorKind.noLocation).primaryRecovery,
        RouteRecovery.retry,
      );
      expect(
        routeErrorView(l10n, RouteErrorKind.unknown).primaryRecovery,
        RouteRecovery.retry,
      );
    });
  });
}

File: apps/web/test/state/route-error.test.ts
// 移植元: lib/core/models/route_error.dart と test/core/models/route_error_test.dart

import { describe, expect, it } from 'vitest';

import {
  ClientException,
  TimeoutException,
} from '@aruku/engine/services/http-client';
import { RouteException } from '@aruku/engine/services/route-service';

import { RouteErrorKind } from '../../src/state/app-state';
import { classifyRouteError, routeErrorView } from '../../src/state/route-error';

describe('エンジンが投げるステータスの分類', () => {
  it.each([
    ['NO_ORIGIN', RouteErrorKind.noLocation],
    ['NO_DESTINATION', RouteErrorKind.noDestination],
    ['ZERO_RESULTS', RouteErrorKind.noResults],
    ['TIMEOUT', RouteErrorKind.timeout],
  ])('%s は %s', (status, expected) => {
    expect(classifyRouteError(new RouteException(status))).toBe(expected);
  });

  it.each(['HTTP 500', 'HTTP 429', 'HTTP 401'])('%s は通信系', (status) => {
    expect(classifyRouteError(new RouteException(status))).toBe(
      RouteErrorKind.network,
    );
  });

  // 設定・配線の失敗。ユーザーの操作では直らないので、再試行を促す network にも
  // 条件変更を促す noResults にも寄せない。
  it.each(['NO_TRANSIT_API', 'MATRIX_NOT_ARRAY'])('%s は原因不明扱い', (status) => {
    expect(classifyRouteError(new RouteException(status))).toBe(
      RouteErrorKind.unknown,
    );
  });
});

describe('例外そのものの分類', () => {
  // タイムアウトを network に丸めない。上流の遅延は通信断ではなく、「通信状況を
  // 確認して再試行」は電波が正常なユーザーを的外れな導線へ送る（#300）。
  it('TimeoutException は遅延であって通信断ではない', () => {
    expect(classifyRouteError(new TimeoutException('header timeout'))).toBe(
      RouteErrorKind.timeout,
    );
  });

  // fetch の TypeError はトランスポート層（FetchHttpClient）が ClientException へ
  // 寄せて届く。移植元が dart:io の IOException と http の ClientException を両方
  // 見ていたのは、dart2js の dart:io がスタブで Web では一致しないため（#359）。
  // こちらは層が1つに揃えているので、その二重性は運ばない。
  it('ClientException は通信系', () => {
    expect(classifyRouteError(new ClientException('Failed to fetch'))).toBe(
      RouteErrorKind.network,
    );
  });

  // ここを network にすると、本物のバグが「通信状況を確認してください」として
  // ユーザーにも我々にも見えなくなる。通信の TypeError は上記のとおり包まれて届く。
  it('素の TypeError は通信系にしない', () => {
    expect(classifyRouteError(new TypeError('x is not a function'))).toBe(
      RouteErrorKind.unknown,
    );
  });

  it('例外ですらない値も落ちずに分類できる', () => {
    expect(classifyRouteError('ごみ')).toBe(RouteErrorKind.unknown);
    expect(classifyRouteError(null)).toBe(RouteErrorKind.unknown);
    expect(classifyRouteError(undefined)).toBe(RouteErrorKind.unknown);
  });
});

describe('復帰導線', () => {
  // 種別ごとに「まず何をさせるか」が変わる。再試行で直らないものに再試行を
  // 出すと、同じ失敗を繰り返させる。
  it.each([
    [RouteErrorKind.network, 'retry'],
    [RouteErrorKind.timeout, 'retry'],
    [RouteErrorKind.noLocation, 'retry'],
    [RouteErrorKind.unknown, 'retry'],
    [RouteErrorKind.noResults, 'changeConditions'],
    [RouteErrorKind.noDestination, 'changeConditions'],
  ])('%s の主導線は %s', (kind, recovery) => {
    expect(routeErrorView(kind).primaryRecovery).toBe(recovery);
  });

  it('種別ごとに違う文言を持つ', () => {
    const kinds = Object.values(RouteErrorKind);
    const titles = kinds.map((k) => routeErrorView(k).title);

    expect(new Set(titles).size).toBe(kinds.length);
    for (const kind of kinds) {
      expect(routeErrorView(kind).title).not.toBe('');
      expect(routeErrorView(kind).description).not.toBe('');
    }
  });
});
